web_src/js/features/user-auth-webauthn.js

```diff
@@ -0,0 +1,196 @@
+      rawId: bufferEncode(rawId),
+      type: newCredential.type,
+      response: {
+        attestationObject: bufferEncode(attestationObject),
+        clientDataJSON: bufferEncode(clientDataJSON),
+      },
+    }),
+    dataType: 'json',
+    contentType: 'application/json; charset=utf-8',
+  }).then(() => {
+    window.location.reload();
+  }).fail((xhr) => {
+    if (xhr.status === 409) {
+      webAuthnError('duplicated');
+      return;
+    }
+    webAuthnError('unknown');
+  });
+}
+
+function webAuthnError(errorType, message) {
+  $('#webauthn-error [data-webauthn-error-msg]').hide();
+  if (errorType === 0 && message && message.length > 1) {
+    $(`#webauthn-error [data-webauthn-error-msg=0]`).text(message);
+    $(`#webauthn-error [data-webauthn-error-msg=0]`).show();
+  } else {
+    $(`#webauthn-error [data-webauthn-error-msg=${errorType}]`).show();
+  }
+  $('#webauthn-error').modal('show');
+}
+
+function detectWebAuthnSupport() {
+  if (!window.isSecureContext) {
+    $('#register-button').prop('disabled', true);
+    $('#login-button').prop('disabled', true);
+    webAuthnError('insecure');
+    return false;
+  }
+
+  if (typeof window.PublicKeyCredential !== 'function') {
+    $('#register-button').prop('disabled', true);
+    $('#login-button').prop('disabled', true);
+    webAuthnError('browser');
+    return false;
+  }
+
+  return true;
+}
+
+export function initUserAuthWebAuthnRegister() {
+  if ($('#register-webauthn').length === 0) {
+    return;
+  }
+
+  if (!detectWebAuthnSupport()) {
+    return;
+  }
+
+  $('#register-device').modal({allowMultiple: false});
+  $('#webauthn-error').modal({allowMultiple: false});
+  $('#register-webauthn').on('click', (e) => {
+    e.preventDefault();
+    webAuthnRegisterRequest();
+  });
+}
+
+function webAuthnRegisterRequest() {
+  if ($('#nickname').val() === '') {
+    webAuthnError('empty');
+    return;
+  }
+  $.post(`${appSubUrl}/user/settings/security/webauthn/request_register`, {
+    _csrf: csrfToken,
+    name: $('#nickname').val(),
+  }).done((makeCredentialOptions) => {
+    $('#nickname').closest('div.field').removeClass('error');
+    $('#register-device').modal('show');
+
+    makeCredentialOptions.publicKey.challenge = decode(makeCredentialOptions.publicKey.challenge);
+    makeCredentialOptions.publicKey.user.id = decode(makeCredentialOptions.publicKey.user.id);
+    if (makeCredentialOptions.publicKey.excludeCredentials) {
+      for (let i = 0; i < makeCredentialOptions.publicKey.excludeCredentials.length; i++) {
+        makeCredentialOptions.publicKey.excludeCredentials[i].id = decode(makeCredentialOptions.publicKey.excludeCredentials[i].id);
+      }
+    }
+
+    navigator.credentials.create({
+      publicKey: makeCredentialOptions.publicKey
+    }).then(webauthnRegistered)
+      .catch((err) => {
+        if (err === undefined) {
+          webAuthnError('unknown');
+          return;
+        }
+        webAuthnError(0, err);
```

Comment:
Here `err` seems to be an `Error` object

Reply:
In `initUserAuthWebAuthn`:

```
        }).catch((err) => {
          webAuthnError(0, err.message);
        });
```

Reply:
When I was testing I originally changed it to use `err.message` but as far as I could see that led to an error being thrown here if there was an error. If you can in your own testing assert that this **is** an error and will **always** have a message you can change it.